ui/lib/components/teams/service/ServiceCredentialForm.js

```
@@ -0,0 +1,333 @@
+        return this.setState({
+          submitting: false
+        })
+      }
+
+      if (!values.namespace) {
+        return this.setState({
+          submitting: false,
+          formErrorMessage: 'Please select the namespace!',
+          validationErrors: null
+        })
+      }
+
+      const cluster = clusters.items.find(c => c.metadata.name === values.cluster)
+      const service = services.items.find(s => s.metadata.name === values.service)
+      const namespaceClaim = namespaceClaims.items.find(n => n.spec.name === values.namespace)
+      const name = values.name
+
+      
+      try {
+        const existing = await (await KoreApi.client()).GetServiceCredentials(this.props.team.metadata.name, name)
+        if (Object.keys(existing).length) {
+          return this.setState({
+            submitting: false,
+            formErrorMessage: `A service credential with the name "${name}" already exists`,
+            validationErrors: null
+          })
+        }
+      } catch(err) {
+        // TODO: we should differentiate between 404 and other errors
+      }
+
+      const spec = {
+        name,
+        kind: service.spec.kind,
+        cluster: {
+          group: cluster.apiVersion.split('/')[0],
+          version: cluster.apiVersion.split('/')[1],
+          kind: cluster.kind,
+          name: cluster.metadata.name,
+          namespace: this.props.team.metadata.name
+        },
+        service: {
+          group: service.apiVersion.split('/')[0],
+          version: service.apiVersion.split('/')[1],
+          kind: service.kind,
+          name: service.metadata.name,
+          namespace: this.props.team.metadata.name
+        },
+        clusterNamespace: namespaceClaim.spec.name,
+        configuration: config,
+      }
+      const serviceCredentialResource = Generic({
+        apiVersion: 'servicecredentials.services.kore.appvia.io/v1alpha1',
+        kind: 'ServiceCredentials',
+        name: name,
+        spec
+      })
```

Generally for new implementations, we've been factoring things like this out into functions which use the auto-generated model classes instead of hard-coding the structures in here, so we get unit test failures if the models change.

See, for example, https://github.com/appvia/kore/blob/master/ui/lib/components/policies/PolicyForm.js#L77-L97

Our intention is to gather all of those functions out into a separate lib which can then be unit tested in a standalone manner, but at the moment they're defined in line where needed.

---

Fixed in e065fe5c6363dff9772f123c9867506290f1512a

---

Fixed in 6c82189e93bda46818285c30f27cc6358f0c7d0a